### PR TITLE
Linear solver implementation for compressible

### PR DESCRIPTION
**Pull request by colinjcotter**
This implements the linear compressible solver for dcore

### CODE REVIEW

File: dcore/linear_solvers.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        n = FacetNormal(mesh)
+        
+        #Analytical elimination of theta
+        theta = -u[2]*state.thetabar*beta + theta_in
+
+        eqn = (
+            (inner(w , u) - beta*div(theta*w)*pibar)*dx
+            + beta*jump(theta*w,n)*avg(pibar)*dS_v
+            - beta*div(thetabar*w)*(pibar_theta*theta + pibar_rho*rho)*dx
+            + beta*jump(thetabar*w,n)*avg(pibar_theta*theta + pibar_rho*rho)*dS
```
Comment (dorugeber) [inline dcore/linear_solvers.py]: `dS` alert

File: dcore/linear_solvers.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
+From firedrake import split, LinearVariationalProblem, \
```
Comment (wence-) [inline dcore/linear_solvers.py]: from, not From.

File: dcore/linear_solvers.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    def __init__(x_in, x_out, state, alpha = 0.5):
+        super(CompressibleSolver, self).__init(x_in, x_out)
+
+        self.state = state
+        self.alpha = alpha
+        
+        #setup the solver
+        self._setup_solver()
+
+    def _setup_solver:
```
Comment (wence-) [inline dcore/linear_solvers.py]: `(self)`

File: dcore/linear_solvers.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    This solver follows the following strategy:
+    (1) Analytically eliminate theta (introduces error near topography)
+    (2) Solve resulting system for (u,rho) using a Schur preconditioner
+    (3) Reconstruct theta
+
+    :arg x_in: :class:`.Function` object for the input
+    :arg x_out: :class:`.Function` object for the output
+    :arg state: a :class:`.State` object containing everything else.
+    :arg alpha: off-centring parameter from [0,1] (default value 0.5).
```
Comment (wence-) [inline dcore/linear_solvers.py]: centering

File: dcore/linear_solvers.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        params={'pc_type': 'fieldsplit',
+                'pc_fieldsplit_type': 'schur',
+                'ksp_type': 'gmres',
+                'ksp_max_it': 100,
+                'ksp_gmres_restart': 50,
+                'pc_fieldsplit_schur_fact_type': 'FULL',
+                'pc_fieldsplit_schur_precondition': 'selfp',
+                'fieldsplit_0_ksp_type': 'preonly',
+            'fieldsplit_0_pc_type': 'bjacobi',
```
Comment (wence-) [inline dcore/linear_solvers.py]: indent

File: dcore/linear_solvers.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        self.theta = Function(state.Vt)
+        
+        theta_eqn = gamma*(theta -u[2]*state.thetabar*beta + theta_in)*dx
+        theta_problem = LinearVariationalProblem(lhs(theta_eqn),
+                                                 rhs(theta_eqn),
+                                                 self.theta)
+        self.theta_solver = LinearVariationalSolver(theta_problem)
+
+        
+    def solve():
```
Comment (wence-) [inline dcore/linear_solvers.py]: `self`

File: dcore/linear_solvers.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+    This is a dummy base class where the input is just copied to the output.
+
+    :arg x_in: :class:`.Function` object for the input
+    :arg x_out: :class:`.Function` object for the output
+    """
+
+    def __init__(x_in, x_out):
+        self.x_in = x_in #get input vector
+        self.x_out = x_out #get output vector
```
Comment (wence-) [inline dcore/linear_solvers.py]: I presume all solvers will need `state`, so why not put it here?

File: dcore/linear_solvers.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    
+    def solve():
+        """
+        Function to execute the solver.
+        """
+        #This is a base class so we just copy x_in to x_out
+        self.x_out.assign(x_in)
```
Comment (wence-) [inline dcore/linear_solvers.py]: Consider abstract base classes?  Which is what this is?

```python
from abc import ABCMeta, abstractmethod
class TimeStepper(object):
    __metaclass__ == ABCMeta
   
    def __init__(self, ...);
        pass

    @abstractmethod
    def solve(self):
        pass
```

Comment (colinjcotter) [inline dcore/linear_solvers.py]: Only thing is, that doesn't document the x_in, x_out requirement.

Comment (wence-) [inline dcore/linear_solvers.py]: Huh, I don't know what you mean.  You can certainly add a docstring.  You can even have a method body which can be called with `super`.  You're just required to override this function in all direct subclasses of `TimeStepper`.

**Comment by colinjcotter** [comment]
Oh, so can I have my 
 self.x_in = x_in #get input vector
        self.x_out = x_out #get output vector

in the __init__ method?

**Comment by wence-** [comment]
Yes, you can use them, in all regards, like a normal class, except that methods marked with `@abstractmethod` *must* be overridden by the child class.

**Comment by colinjcotter** [comment]
better now?

File: dcore/linear_solvers.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    (3) Reconstruct theta
+
+    :arg x_in: :class:`.Function` object for the input
+    :arg x_out: :class:`.Function` object for the output
+    :arg state: a :class:`.State` object containing everything else.
+    :arg alpha: off-centering parameter from [0,1] (default value 0.5).
+    """
+
+    def __init__(self, x_in, x_out, state, alpha = 0.5):
+        super(CompressibleSolver, self).__init(x_in, x_out)
```
Comment (wence-) [inline dcore/linear_solvers.py]: `__init__`

**Pull request merged**